### ISSUE TITLE
precision over 1000 causing DecimalError when formatting a BigNumber

### ISSUE THREAD

**Issue by husayt** [josdejong/mathjs#1385]
When I try big precisions over 1000 this is the error I get:

```
: Error. [DecimalError] Precision limit exceededA
```

If there is a limit to precision, would be nice to document this, otherwise this is an issue

**Comment by josdejong** [comment]
According to the docs of Decimal.js:

> The number of digits of value is not limited, except by JavaScript's maximum array size and, in practice, the processing time required.

Looking in the code of decimal.js though, I see that `pi` is a hardcoded with 1025 digits, and there is an error thrown if you request pi with a higher precision than that:

https://github.com/MikeMcl/decimal.js/blob/master/decimal.js#L33
https://github.com/MikeMcl/decimal.js/blob/master/decimal.js#L3158-L3161

To resolve this I guess we could fallback on calculating pi on the fly with higher precision.

**Comment by josdejong** [comment]
Note that calculations without pi can be done with higher precision without any issue:

```js
math.config({ precision: 10000 })

math.bignumber(1).div(3).toString() // 0.3333333... many digits...
```

**Comment by husayt** [comment]
Why then 1/3 gives me that error when precision is over 1000?

**Comment by harrysarson** [comment]
Could you share an example of code that throws that error?

**Comment by husayt** [comment]
Yes sure:
https://codesandbox.io/s/9240x1xqzp

**Comment by josdejong** [comment]
That's interesting. Looks indeed that `math.format` triggers the limit on 1025 that Decimal.js has. Would be great if anyone can look into the cause and possible workaround for this. 

On a side note: I think working with a precision over 1000 digits is theoretical? Or is there a real daily use case for this?

**Comment by husayt** [comment]
@ericman314 Thanks for a solution and detailed explanation.

**Comment by josdejong** [comment]
@husayt is the fix of Eric enough to solve your issue? It means that precision above 1000 digits works for arithmetic operations like multiplication and addition, but still not for trigonometric operations involving `pi`.

**Comment by husayt** [comment]
Yes this issue with formatting is thankfully resolved, and yes Pi issue is
still on the table and hopefully will find its solution soon. They are
related, but separate issues. I noticed in discussion you had a good idea
about resolving that too, so looking forward to it .

Thanks

**Comment by josdejong** [comment]
:+1: then I will keep this issue open as a reminder

**Comment by josdejong** [comment]
The issue with `math.format` is now fixed in `v5.4.2` thanks to Eric's fix.

**Comment by MikeMcl** [comment]
See my answer to [this](https://stackoverflow.com/questions/46744170/precision-issue-with-decimal-js-tan/48243281#48243281) stackoverlow question for more details on this limitation.
